### PR TITLE
KeyBinding and saveLoadCameraPosition fixes #411 fixes #410

### LINKED ISSUE #410: Key Bindings
Want to support a mapping of keys to actions.  This should be configurable and use the existing Veldrid Key implementations.

This ticket was born from adding Camera save/load where it was discovered that ZH had the capability on the f1-f8 keys (with ctrl+ to save).  Given that OSage uses the f2-4 keys for debugging it makes sense to abstract the responsibility of Key-bindings into its own controller and have that controller load keybinding files (with one generated by default if it doesn't exist).

### LINKED ISSUE #411: SaveLoad Camera Position
In ZH ctrl+[f1-f8] saves the camera position and can be loaded with [f1-f8] (Thanks Han).  In RtsCameraController add logic to store/load (in memory) 'n' number of camera positions (no reason to limit the number of camera-recalls as long as users can add enough hotkeys).  I am currently working toward a solution here.

### PR DESCRIPTION
**Pull request by chemdude8**
Still looking to update unit tests but wanted to put this out there while I have a question in the discord.

### CODE REVIEW

**Comment by chemdude8** [comment]
fixes #411 fixes #410

File: src/OpenSage.Game/Graphics/Cameras/CameraLookData.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
+using System.Numerics;
+using System.Text;
+
+namespace OpenSage.Graphics.Cameras
+{
+
+    class CameraLookData
+    {
+        public Vector3 newPosition { get; set; }
```
Comment (feliwir) [inline src/OpenSage.Game/Graphics/Cameras/CameraLookData.cs:9]: Must be `NewPosition`

File: src/OpenSage.Game/Graphics/Cameras/CameraLookData.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+        public float pitch { get; set; }
+
+        public float zoom { get; set; }
+
+        public float intersect { get; set; }
+
+        public override string ToString()
+        {
+            StringBuilder sb = new StringBuilder();
```
Comment (feliwir) [inline src/OpenSage.Game/Graphics/Cameras/CameraLookData.cs:25]: We usually use `var` on the left side when possible

File: src/OpenSage.Game/Graphics/Cameras/CameraLookData.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            foreach (System.Reflection.PropertyInfo property in this.GetType().GetProperties())
```
Comment (feliwir) [inline src/OpenSage.Game/Graphics/Cameras/CameraLookData.cs:26]: Use `var` here aswell

File: src/OpenSage.Game/Graphics/Cameras/RtsCameraController.cs [hunk 4]
```diff
@@ -204,6 +210,23 @@ void ICameraController.UpdateCamera(Camera camera, in CameraInputState inputStat
...
                 Vector3.UnitZ);
+
+            CameraLookData lookData = createLookData(newPosition, targetPosition);
+            checkForCameraPositionSave(inputState, lookData);
+            checkForCameraPositionLoad(inputState, lookData, camera);
+        }
+
+        private CameraLookData createLookData(Vector3 newPosition, Vector3 targetPosition)
+        {
+            CameraLookData lookData = new CameraLookData();
```
Comment (feliwir) [inline src/OpenSage.Game/Graphics/Cameras/RtsCameraController.cs:221]: I'd suggest an object initializer: https://docs.microsoft.com/de-de/dotnet/csharp/programming-guide/classes-and-structs/how-to-initialize-objects-by-using-an-object-initializer

Also use var here again

File: src/OpenSage.Game/Input/KeyBinding/KeyBindingController.cs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+                return null;
+            }
+
+            if (_keyBindings == null)
+            {
+                Logger.Warn("Failed to find keybindings, key actions will not be enabled.");
+                return null;
+            }
+
+            foreach (KeyBinding binding in _keyBindings)
```
Comment (feliwir) [inline src/OpenSage.Game/Input/KeyBinding/KeyBindingController.cs:64]: Use `var`

File: src/OpenSage.Game/Input/KeyBinding/KeyBindingController.cs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+            {
+                if (binding.keyAction.Equals(action))
+                {
+                    bool missingKey = false;
+                    foreach (Key bindingKey in binding.keys)
```
Comment (feliwir) [inline src/OpenSage.Game/Input/KeyBinding/KeyBindingController.cs:69]: Use `var`

**Comment by tgjones** [comment]
Firstly, thank you for taking the time to put this PR together.

I'm going to close it though, because we'll want the key binding system use the existing `CommandMap.ini` file that comes with the SAGE games, which already binds keys to actions, i.e.

```
CommandMap SAVE_VIEW1
  Key = KEY_F1
  Transition = DOWN
  Modifiers = CTRL
  UseableIn = GAME
  Category = INTERFACE
  Description = GUI:SaveView1Description
  DisplayName = GUI:SaveView1
End
```

The `CameraLookData` class in this PR is roughly what we will want, so this will be a handy reference for building the `CommandMap.ini`-based solution, thank you.

**Pull request closed without merging**